internal/controllers/core/cluster/monitor.go

```diff
@@ -3,33 +3,98 @@ package cluster
 import (
 	"context"
 	"errors"
+	"sync"
 
+	"github.com/jonboulle/clockwork"
 	"k8s.io/apimachinery/pkg/types"
 
+	"github.com/tilt-dev/tilt/internal/controllers/indexer"
 	"github.com/tilt-dev/tilt/internal/k8s"
 )
 
-func (r *Reconciler) monitorConn(ctx context.Context, clusterNN types.NamespacedName, conn connection) {
+type clusterHealthMonitor struct {
+	mu        sync.Mutex
+	globalCtx context.Context
+	clock     clockwork.Clock
+	requeuer  *indexer.Requeuer
+	monitors  map[types.NamespacedName]monitor
+}
+
+func newClusterHealthMonitor(globalCtx context.Context, clock clockwork.Clock, requeuer *indexer.Requeuer) *clusterHealthMonitor {
+	return &clusterHealthMonitor{
+		globalCtx: globalCtx,
+		clock:     clock,
+		requeuer:  requeuer,
+		monitors:  make(map[types.NamespacedName]monitor),
+	}
+}
+
+func (c *clusterHealthMonitor) Start(clusterNN types.NamespacedName, conn connection) context.Context {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.cleanup(clusterNN)
+	ctx, cancel := context.WithCancel(c.globalCtx)
+	c.monitors[clusterNN] = monitor{cancel: cancel}
+	go c.run(ctx, clusterNN, conn)
+
+	return ctx
+}
+
+func (c *clusterHealthMonitor) GetStatus(clusterNN types.NamespacedName) string {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.monitors[clusterNN].error
+}
+
+func (c *clusterHealthMonitor) UpdateStatus(clusterNN types.NamespacedName, error string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if m, ok := c.monitors[clusterNN]; ok {
+		if m.error == error {
+			return
+		}
+		m.error = error
+		c.monitors[clusterNN] = m
+		c.requeuer.Add(clusterNN)
+	}
+}
+
+func (c *clusterHealthMonitor) Stop(clusterNN types.NamespacedName) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.cleanup(clusterNN)
+	delete(c.monitors, clusterNN)
+}
+
+func (c *clusterHealthMonitor) cleanup(clusterNN types.NamespacedName) {
+	m := c.monitors[clusterNN]
+	if m.cancel != nil {
+		m.cancel()
```

Comment:
i think you have a slight race condition here -- `cancel()` assumes that it stops the healthcheck loop immediately, but in your code, the loop can keep going and recording results.

there's two patterns that we usually do to deal with this:
1) Add some sort of done channel so that your healthcheck loop can tell everyone when it's done. (i.e.,have the controller code synchronize before starting a new loop)
2) Change UpdateStatus() to do nothing if ctx.Err() != nil (i.e., have the healthcheck loop synchronize before sending data out to the controller)

Reply:
alternatively, the healthcheck loop could have a pointer to its own monitor and only update that one; creating a new loop creates a new monitor

Reply:
Nice catch - went with (2) in [dbadc7f](https://github.com/tilt-dev/tilt/pull/5781/commits/dbadc7fa5d37b29573764f3f2322a0703d1a0023) as it was the most straightforward for now!